### DISCUSSION TITLE
Build Error devicetree with umlaut

### DISCUSSION THREAD

**Discussion by CHSchuepfer** [nickcoutsos/keymap-editor#181] [Q&A] [answered]
Hey everyone,

im about to get my Advantage 360 (non Pro). So being a good german and wanting  to check off a stereotype i decied to prepare, meaning reading up and play around with the whole Keymapping situation.

I stumpled upon this repo and your website. In terms of workflow and convinience its really good.
The next step was checking how the situation was with Umlaute (ö,ä,ü and so on).

I then saw the discussion here https://github.com/nickcoutsos/keymap-editor/discussions/146 and the issue here https://github.com/nickcoutsos/keymap-editor/issues/166.

I added the keys_de.h in the `adv360.keymap` file and could build the firmware with the blank/default mappings.
Today i tried to set them up as i would like them to have (coming from a moonlander).

But now i run into build errors:

```log
Run west build -s zmk/app -d build/left -b adv360_left -- -DZMK_CONFIG="${GITHUB_WORKSPACE}/config"
-- west build: generating a build system
-- Adding ZMK config directory as board root: /__w/zmk-config-adv360/zmk-config-adv360/config
-- ZMK Config directory: /__w/zmk-config-adv360/zmk-config-adv360/config
-- Using keymap file: /__w/zmk-config-adv360/zmk-config-adv360/config/adv360_left.keymap
Including boilerplate (Zephyr base): /__w/zmk-config-adv360/zmk-config-adv360/zephyr/cmake/app/boilerplate.cmake
-- Application: /__w/zmk-config-adv360/zmk-config-adv360/zmk/app
-- Zephyr version: 2.5.0 (/__w/zmk-config-adv360/zmk-config-adv360/zephyr)
-- Found Python3: /usr/bin/python3.7 (found suitable exact version "3.7.3") found components: Interpreter 
-- Found west (found suitable version "0.8.0", minimum required is "0.7.1")
-- git describe failed: fatal: No tags can describe 'f2108f16[4](https://github.com/CHSchuepfer/zmk-config-adv360/actions/runs/7776851300/job/21204474729#step:8:5)6b964adffb9c7ed74641c828fef2a44'.
Try --always, or create some tags.;
   BUILD_VERSION is left undefined
-- Board: adv360_left
-- Cache files will be written to: /__w/zmk-config-adv360/zmk-config-adv360/zephyr/.cache
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Using toolchain: zephyr 0.11.4 (/opt/zephyr-sdk-0.11.4)
-- Found dtc: /opt/zephyr-sdk-0.11.4/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.[5](https://github.com/CHSchuepfer/zmk-config-adv360/actions/runs/7776851300/job/21204474729#step:8:6).0", minimum required is "1.4.[6](https://github.com/CHSchuepfer/zmk-config-adv360/actions/runs/7776851300/job/21204474729#step:8:7)")
-- Found BOARD.dts: /__w/zmk-config-adv360/zmk-config-adv360/config/boards/arm/adv360/adv360_left.dts
-- Found devicetree overlay: /__w/zmk-config-adv360/zmk-config-adv360/config/adv360_left.keymap
Error: adv360_left.dts.pre.tmp:800.405-406 syntax error
FATAL ERROR: Unable to parse input tree
CMake Error at /__w/zmk-config-adv360/zmk-config-adv360/zephyr/cmake/dts.cmake:205 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /__w/zmk-config-adv360/zmk-config-adv360/zephyr/cmake/app/boilerplate.cmake:535 (include)
  /__w/zmk-config-adv360/zmk-config-adv360/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:24 (include)
  /__w/zmk-config-adv360/zmk-config-adv360/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:35 (include_boilerplate)
  CMakeLists.txt:1[7](https://github.com/CHSchuepfer/zmk-config-adv360/actions/runs/7776851300/job/21204474729#step:8:8) (find_package)


-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /usr/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/__w/zmk-config-adv360/zmk-config-adv360/build/left -S/__w/zmk-config-adv360/zmk-config-adv360/zmk/app -GNinja -DBOARD=adv360_left -DZMK_CONFIG=/__w/zmk-config-adv360/zmk-config-adv360/config
Error: Process completed with exit code 1.
```

I didnt change anything except the actual mapping.
The Error gets thrown in the `Web Build (left)` Build Step.

My Keymap Repo is public here: https://github.com/CHSchuepfer/zmk-config-adv360

Can someone please take a look at it and help me find where my mistake is, because for the life of me i cant find it.

Thanks in advance :).

**Comment by nickcoutsos** [comment, marked as answer]
Hi!

This is... interesting. This build error tells you

> Error: adv360_left.dts.pre.tmp:800.405-406 syntax error

And if you look at that section of the DTS generated by the build process you'll see (formatted for readability)

```dts
&kp ((((0x07) << 16) | (0x2B)))
&kp ((((0x07) << 16) | (0x14)))
&kp ((((0x07) << 16) | (0x1A)))
&kp ((((0x07) << 16) | (0x08)))
&kp ((((0x07) << 16) | (0x15)))
&kp ((((0x07) << 16) | (0x17)))
&kp ((((0x07) << 16) | (0x4D))) &macro_kinesis
&kp ((((0x07) << 16) | (0x1D)))
&kp ((((0x07) << 16) | (0x18)))
&kp ((((0x07) << 16) | (0x0C)))
&kp ((((0x07) << 16) | (0x12)))
&kp ((((0x07) << 16) | (0x13)))
&kp (ZMK_HID_USAGE((0x07), (0x2F)))
```

The column range points to that last part, specifically the start of `ZMK_HID_USAGE`. At first I thought maybe, weirdly, the header order was causing it to not evaluate that pre-processor macro but that didn't help. I noticed that you're building against a branch of ZMK that's a couple of years old. Sometime since that branch there were some macro functions that have been renamed:

- Old (Refil/zmk/adv360-beta): https://github.com/ReFil/zmk/blob/adv360-beta/app/include/dt-bindings/zmk/hid_usage_pages.h
- Current (zmkfirmware/zmk/main): https://github.com/zmkfirmware/zmk/blob/main/app/include/dt-bindings/zmk/hid_usage_pages.h

For a _minimal_ lift you _may_ be able to get away with copying the missing macros (https://github.com/zmkfirmware/zmk/blob/5826b80374625d448cfbfc739dde4fda1e6f2681/app/include/dt-bindings/zmk/hid_usage_pages.h#L13-L15) into your keymap or into `keys_de.h` -- I'm not sure what else may be different.

You would benefit from switching to something more recent for other reasons. ZMK has changed a fair bit, sometimes breaking changes, and for simplicity Keymap Editor will often defer to modern ZMK's expectations. I don't think it should be too bad in your case, possibly:

1. Fork from [Kinesis' repo](https://github.com/KinesisCorporation/Adv360-Pro-ZMK)
2. Copy your existing keymap into [ZMK's keymap upgrader](https://zmk.dev/keymap-upgrader)
3. Add  `keys_de.h` to the new repo

**Reply by CHSchuepfer** [reply]
Hey, thanks for the reply!
You could be right. I forked the repo from https://github.com/PolarityWorks/zmk-config-adv360 because i didnt know i could use the ZMK 360 Pro Profile for the 360. If its safe to use then i of course will do that. Just wasnt sure because of the bluetooth drivers which are probl. missing or not necesarry on the non pro modell.

Rebuilding the repo isnt a big deal, i just kill the current one and fork fresh in a separate repo.

Just for me to be sure (bc i dont have the hardware yet and just wanted to familiarize myself with the structure):

The right and left keymap files get generated out of the overall keymap file. Which is why i add the `keys_de.h` files to the adv360.keymap file, right?

I will rebuild the repo, try again and report back.
Thanks again for the reply :)


So long

**Reply by CHSchuepfer** [reply]
Ok so i tested it and with the few changes i made and the build runs succesfull.
So you might be right on the money that i just, stupidly, used a 2 yo Firmware :laughing: 

Device should arrive today or tomorrow, i can then "go live" so to speak.
Just hope it works and i can use the 360 Pro firmware on the 360 (crossing fingers as to not brick my device) :rofl: 


Thanks again for the help :)

**Reply by nickcoutsos** [reply]
Hey... you may be better off chatting with Kinesis, but from what I know the 360 can't run ZMK.

That's not a question of using the PolarityWorks repo vs the Kinesis one, or of whether or not Bluetooth is available. I don't know what microcontroller the non-pro version is using but I haven't heard anybody talk about flashing any firmware on it. You'll want to look at their SmartSet software to handle remapping.

**Reply by CHSchuepfer** [reply]
You are correct, ZMK isnt available on the non Pro.
Even though it was in vain thanks again for the help.

so long
